base/shared/src/main/scala/scalaz/data/io.scala

```diff
@@ -0,0 +1,109 @@
+package scalaz.data
+
+import scalaz._
+
+import scala.Option
+import scala.Either
+import scala.collection.generic.CanBuildFrom
+import scala.collection.mutable
+import scala.inline
+
+import scalaz.zio.Errors.NothingRaced
+import scalaz.zio.{ Fiber, IO, Infallible, Ref }
+
+trait IOFunctions {
+
+  final def absolve[E, A](v: IO[E, E \/ A]): IO[E, A] =
+    v.flatMap(fromEither)
+
+  final def fromEither[E, A](v: E \/ A): IO[E, A] =
+    v.fold(IO.fail, IO.now)
+
+  // TODO check if this can be optimized
+  private final def _bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, Maybe[E \/ B]) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] = {
+    val release1: (A, Option[Either[E, B]]) => Infallible[Unit] =
+      (a: A, o: Option[Either[E, B]]) => release(a, Maybe.fromOption(o.map(Disjunction.fromEither[E, B])))
+    IO.bracket0[E, A, B](acquire)(release1)(use)
+  }
+
+  @inline
+  final def bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, Maybe[E \/ B]) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
+    _bracket0[E, A, B](acquire)(release)(use)
+
+  final def require[E, A](error: E): IO[E, Maybe[A]] => IO[E, A] =
+    (io: IO[E, Maybe[A]]) =>
+      io.flatMap { maybe =>
+        Maybe.fold[A, IO[E, A]](maybe)(IO.now[E, A], IO.fail[E, A](error))
+    }
+
+  final def forkAll[E, A, F[_]: Foldable](
+    as: F[IO[E, A]]
+  )(implicit cbf: CanBuildFrom[F[IO[E, A]], A, F[A]]): IO[E, Fiber[E, F[A]]] =
+    implicitly[Foldable[F]]
+      .foldRight(as, IO.point[E, Fiber[E, mutable.Builder[A, F[A]]]](Fiber.point(cbf(as)))) {
+        case (a, as) => as.par(a.fork).map { case (as, a) => as.zipWith(a)(_ += _) }
+      }
+      .map(as => as.zipWith(Fiber.point(())) { case (as, _) => as.result })
+
+  final def traverse[E, A, B, F[_]: Foldable](
+    in: F[A]
+  )(fn: A => IO[E, B])(implicit cbf: CanBuildFrom[F[A], B, F[B]]): IO[E, F[B]] =
+    implicitly[Foldable[F]]
+      .foldLeft(in, IO.point[E, mutable.Builder[B, F[B]]](cbf(in)))((io, b) => io.zipWith(fn(b))(_ += _))
+      .map(_.result())
+
+  def parTraverse[E, A, B, F[_]: Foldable](
+    in: F[A]
+  )(fn: A => IO[E, B])(implicit cbf: CanBuildFrom[F[A], B, F[B]]): IO[E, F[B]] =
+    implicitly[Foldable[F]]
+      .foldLeft(in, Ref[E, mutable.Builder[B, F[B]]](cbf(in))) {
+        case (ioRef, a) =>
+          ioRef.par(fn(a)).flatMap {
+            case (ref, b) => ref.modify(_ += b) *> IO.point(ref)
+          }
+      }
+      .flatMap(_.read)
+      .map(_.result())
```

Comment:
Hi @sideeffffect , I noticed your PR yesterday and I had the chance to play with it for a while, and
I wasn't able to proof the incorrectness of your implementation :)

If your assumptions hold, I'll definitely amend my implementation as well, removing `Ref` and cutting off some complexity (not to mention performance..).
Will also need to fix the build, since scalaz-zio apis are moving, but that's another story.